### DISCUSSION TITLE
PIP-Compile cannot refresh lock files when private repository

### DISCUSSION THREAD

**Discussion by pan-daniel** [renovatebot/renovate#21838] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I would like that `pip-compile` can match the `--extra-index-url` specified in `requirements.txt` and use the defined `hostRules` to provide it with token/username and encrypted password.

So for example, if the normal defined url is

`---extra-index-url https://url/python/simple/`

based upon my `hostRules` which specify username and password, I'd like it to be changed automatically to

`---extra-index-url https://username:password@url/python/simple/`

Without `--extra-index-url`, the normal update of `requirements.in` works, even for dependencies that are in a private repository, as the `hostRules` get taken into account. The `lockRefresh`, however, fails.

### What you tried so far.

Hello!

Maybe I will find some help here.

While I know that pip-compile is not production ready, I was tempted to try it out after a voice from one of the team members that it is exactly what we want. And to my surprise, everything was going quite smoothly until the pip-compile-refresh phase.

The updating of the requirements.in works just fine when I have the `hostRules` defined with username and encrypted password. The same doesn't work for lock file refresh, so I tried to add the `--extra-index-url`.

Unfortunately, I do not want to expose my token, so the url is just that - an url without a password and username.

But the 'requirements.txt' never update, even though locally with the same command it goes in just fine (obviously I have to give it username and password). It just pumps out this error here, that it aborts, and the only piece of information is User for europe-west3-python.pkg.dev: .

My suspicion is that the pip-compile-refresh fails because it doesn't get the username and password that I set for this particular host in `hostRules`. However, as I mentioned earlier, I have only input the -`-extra-index-url`, because the `registryUrls` that I have defined in my renovatebot config is ignored, or rather, when refreshing the `requirements.txt` it is not taken into account. I would expect that if I have `registryUrls` defined, it works both for updating the `.in` and `.txt` files. Currently it seems to be working for `.in` files only, and comically after adding the `--extra-index-url` that breaks the normal `pip-compile` which IS able somehow to resolve my dependencies from my private repository.

So I do not know - ideally I could drop the `--extra-index-url` altogether from my `requirements.in`, after all I already have defined `hostRules` and `registryUrls` in my config for the packages I really want it to go to my private repository. But if I cannot drop it, I would expect that `pip-compile` gets the token or username and encrypted password from `hostRules` anyway.

Any help here would be appreciated.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: rawExec err(branch="renovate/pip-compile-refresh")
{
  "err": {
    "cmd": "/bin/sh -c docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/company-name/project-name\":\"/mnt/renovate/gh/company-name/project-name\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/company-name/project-name/requirements\" ghcr.io/containerbase/sidecar bash -l -c \"install-tool python 3.10.11 && pip install --user pip-tools && pip-compile --resolver=backtracking requirements.in\"",
    "stderr": "/opt/buildpack/tools/python/3.10.11/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.\n  warnings.warn(\"Setuptools is replacing distutils.\")\n\nAborted!\n",
    "stdout": "installing v2 tool python v3.10.11\nlinking tool python v3.10.11\nPython 3.10.11\npip 23.1.2 from /opt/buildpack/tools/python/3.10.11/lib/python3.10/site-packages/pip (python 3.10)\nInstalled v2 /usr/local/buildpack/tools/v2/python.sh in 8 seconds\nCollecting pip-tools\n  Downloading pip_tools-6.13.0-py3-none-any.whl (53 kB)\n     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.2/53.2 kB 4.4 MB/s eta 0:00:00\nCollecting build (from pip-tools)\n  Downloading build-0.10.0-py3-none-any.whl (17 kB)\nCollecting click>=8 (from pip-tools)\n  Downloading click-8.1.3-py3-none-any.whl (96 kB)\n     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 96.6/96.6 kB 9.1 MB/s eta 0:00:00\nRequirement already satisfied: pip>=22.2 in /opt/buildpack/tools/python/3.10.11/lib/python3.10/site-packages (from pip-tools) (23.1.2)\nRequirement already satisfied: setuptools in /opt/buildpack/tools/python/3.10.11/lib/python3.10/site-packages (from pip-tools) (65.5.0)\nCollecting wheel (from pip-tools)\n  Downloading wheel-0.40.0-py3-none-any.whl (64 kB)\n     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.5/64.5 kB 15.3 MB/s eta 0:00:00\nCollecting packaging>=19.0 (from build->pip-tools)\n  Downloading packaging-23.1-py3-none-any.whl (48 kB)\n     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.9/48.9 kB 11.1 MB/s eta 0:00:00\nCollecting pyproject_hooks (from build->pip-tools)\n  Downloading pyproject_hooks-1.0.0-py3-none-any.whl (9.3 kB)\nCollecting tomli>=1.1.0 (from build->pip-tools)\n  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)\nInstalling collected packages: wheel, tomli, packaging, click, pyproject_hooks, build, pip-tools\nSuccessfully installed build-0.10.0 click-8.1.3 packaging-23.1 pip-tools-6.13.0 pyproject_hooks-1.0.0 tomli-2.0.1 wheel-0.40.0\nUser for europe-west3-python.pkg.dev: ",
    "options": {,
      "encoding": "utf-8",
      "env": {
        "PIP_CACHE_DIR": "/tmp/renovate-cache/others/pip",
        "HOME": "/home/ubuntu",
        "PATH": "/home/ubuntu/.local/bin:/home/ubuntu/bin:/opt/buildpack/tools/python/3.9.3/bin:/home/ubuntu/.npm-global/bin:/home/ubuntu/renovateapp/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
        "LC_ALL": "C.UTF-8",
        "LANG": "C.UTF-8",
        "BUILDPACK_CACHE_DIR": "/tmp/containerbase",
        "CONTAINERBASE_CACHE_DIR": "/tmp/containerbase"
      },
      "maxBuffer": 10485760,
      "timeout": 900000
    },
    "exitCode": 1,
    "name": "ExecError",
    "message": "Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/company-name/project-name\":\"/mnt/renovate/gh/company-name/project-name\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/company-name/project-name/requirements\" ghcr.io/containerbase/sidecar bash -l -c \"install-tool python 3.10.11 && pip install --user pip-tools && pip-compile --resolver=backtracking requirements.in\"\n/opt/buildpack/tools/python/3.10.11/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.\n  warnings.warn(\"Setuptools is replacing distutils.\")\n\nAborted!\n",
    "stack": "ExecError: Command failed: docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/company-name/project-name\":\"/mnt/renovate/gh/company-name/project-name\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e PIP_CACHE_DIR -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/company-name/project-name/requirements\" ghcr.io/containerbase/sidecar bash -l -c \"install-tool python 3.10.11 && pip install --user pip-tools && pip-compile --resolver=backtracking requirements.in\"\n/opt/buildpack/tools/python/3.10.11/lib/python3.10/site-packages/_distutils_hack/__init__.py:33: UserWarning: Setuptools is replacing distutils.\n  warnings.warn(\"Setuptools is replacing distutils.\")\n\nAborted!\n\n    at ChildProcess.<anonymous> (/home/ubuntu/renovateapp/node_modules/renovate/dist/util/exec/common.js:87:24)\n    at ChildProcess.emit (node:events:525:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)"
  },
  "durationMs": 12361
}
```

</details>

**Comment by mbudnek** [comment]
I'm having a very similar issue, but `--extra-index-url` won't really work in my case, since we don't keep that in our `requirements.in`.  Instead our developers have the extra index URL configured globally in our pip.conf.

**Comment by wbolster** [comment]
(not offering a solution here, just some thoughts)

it may be useful to know that the env vars `PIP_INDEX_URL` and `PIP_EXTRA_INDEX_URL` can be used instead of `--index-url` and `--extra-index-url` command line flags. perhaps these can be set/injected somewhere.

**Reply by viceice** [reply]
yes, they can be added via custom env vars config option

https://docs.renovatebot.com/self-hosted-configuration/#customenvvariables

**Reply by RHeynsZa** [reply]
Seems like pip-compile doesnt respect these

**Comment by pan-daniel** [comment]
Just to add my two cents here, as I didn't expect it to get any traction since the creation of this discussion in April.

We cannot really be adding our own env vars, since we opted for using the non-self-hosted version.

We ,,solved" that by migrating to Poetry, which comes with its own set of hurdles, but the needed functionality is in there.

**Comment by RHeynsZa** [comment]
Still facing this, to this day. 

pip-compile doesnt seem to respect `PIP_INDEX_URL` and `PIP_EXTRA_INDEX_URL`

Renovate runs fine, sees updates from the private repo. But as soon as it tries to refresh lock files, it fails with 404 on packages. (because they are private)

**Reply by mbudnek** [reply]
Add `--extra-index-url https://your.private/repo` to your requirements.in file.  Renovate will extract that flag and look up the correct credentials in its `hostRules`.  See https://docs.renovatebot.com/getting-started/private-packages/#pip-compile
